osu.Game.Rulesets.Osu.Tests/Mods/TestSceneOsuModSpunOut.cs

```diff
@@ -48,11 +49,43 @@ public void TestSpinRateUnaffectedByMods(Type additionalModType)
                 PassCondition = () =>
                 {
                     var counter = Player.ChildrenOfType<SpinnerSpmCalculator>().SingleOrDefault();
-                    return counter != null && Precision.AlmostEquals(counter.Result.Value, 286, 1);
+                    var spinner = Player.ChildrenOfType<DrawableSpinner>().FirstOrDefault();
+
+                    if (counter == null || spinner == null)
+                        return false;
+
+                    // ignore cases where the spinner hasn't started as these lead to false-positives
+                    if (Precision.AlmostEquals(counter.Result.Value, 0, 1))
+                        return false;
+
+                    float rotationSpeed = (float)(1.01 * spinner.HitObject.SpinsRequired / spinner.HitObject.Duration);
+
+                    return Precision.AlmostEquals(counter.Result.Value, rotationSpeed * 1000 * 60, 1);
                 }
             });
         }
 
+        [Test]
+        public void TestSpinnerOnlyComplete() => CreateModTest(new ModTestData
+        {
+            Mod = new OsuModSpunOut(),
+            Autoplay = false,
+            Beatmap = singleSpinnerBeatmap,
+            PassCondition = () =>
+            {
+                var spinner = Player.ChildrenOfType<DrawableSpinner>().SingleOrDefault();
+                var gameplayClockContainer = Player.ChildrenOfType<GameplayClockContainer>().SingleOrDefault();
+
+                if (spinner == null || gameplayClockContainer == null)
+                    return false;
+
+                if (!Precision.AlmostEquals(gameplayClockContainer.CurrentTime, spinner.HitObject.StartTime + spinner.HitObject.Duration, 200.0f))
+                    return false;
+
+                return Precision.AlmostEquals(spinner.Progress, 1.0f, 0.05f) && Precision.AlmostEquals(spinner.GainedBonus.Value, 0, 1);
```

Comment:
This feels much cleaner, however I'm noticing an issue.

I've updated the test to use `spinner.IsHit` which makes more sense than using the gameplay clock. I tried looking for something like `spinner.IsCompleted`, however this is the closest thing I could find. The problem is that the spinner unloads sometimes before we can get a judgement reading, so the test fails in those cases.

With the following patch applied (which includes the updated test), the log output shows the following on a "failing" test:
```patch
diff --git a/osu.Game.Rulesets.Osu.Tests/Mods/TestSceneOsuModSpunOut.cs b/osu.Game.Rulesets.Osu.Tests/Mods/TestSceneOsuModSpunOut.cs
index e71377a505..d64f483589 100644
--- a/osu.Game.Rulesets.Osu.Tests/Mods/TestSceneOsuModSpunOut.cs
+++ b/osu.Game.Rulesets.Osu.Tests/Mods/TestSceneOsuModSpunOut.cs
@@ -5,6 +5,7 @@
 using System.Collections.Generic;
 using System.Linq;
 using NUnit.Framework;
+using osu.Framework.Logging;
 using osu.Framework.Testing;
 using osu.Framework.Utils;
 using osu.Game.Beatmaps;
@@ -14,7 +15,6 @@
 using osu.Game.Rulesets.Osu.Objects;
 using osu.Game.Rulesets.Osu.Objects.Drawables;
 using osu.Game.Rulesets.Osu.Skinning.Default;
-using osu.Game.Screens.Play;
 using osuTK;
 
 namespace osu.Game.Rulesets.Osu.Tests.Mods
@@ -66,7 +62,7 @@ public void TestSpinRateUnaffectedByMods(Type additionalModType)
         }
 
         [Test]
-        public void TestSpinnerOnlyComplete() => CreateModTest(new ModTestData
+        public void TestSpinnerGetsNoBonusScore() => CreateModTest(new ModTestData
         {
             Mod = new OsuModSpunOut(),
             Autoplay = false,
@@ -74,15 +70,18 @@ public void TestSpinRateUnaffectedByMods(Type additionalModType)
             PassCondition = () =>
             {
                 var spinner = Player.ChildrenOfType<DrawableSpinner>().SingleOrDefault();
-                var gameplayClockContainer = Player.ChildrenOfType<GameplayClockContainer>().SingleOrDefault();
 
-                if (spinner == null || gameplayClockContainer == null)
+                if (spinner == null)
+                {
+                    Logger.Log("Spinner is null!");
                     return false;
+                }
 
-                if (!Precision.AlmostEquals(gameplayClockContainer.CurrentTime, spinner.HitObject.StartTime + spinner.HitObject.Duration, 200.0f))
-                    return false;
+                Logger.Log(spinner.IsHit.ToString());
+                Logger.Log(Precision.AlmostEquals(spinner.GainedBonus.Value, 0, 1).ToString());
+                Logger.Log("------------------------------");
 
-                return Precision.AlmostEquals(spinner.Progress, 1.0f, 0.05f) && Precision.AlmostEquals(spinner.GainedBonus.Value, 0, 1);
+                return spinner.IsHit && Precision.AlmostEquals(spinner.GainedBonus.Value, 0, 1);
             }
         });
```
![image](https://user-images.githubusercontent.com/68568464/154806182-bcd76418-c759-46b5-a22a-f77d41f3117b.png)

Out of 50 tests, this happens ~15% of the time.
In the visual test, the spinner completes every time, but the test still fails sometimes. If there's any way to get around this race condition, please let me know because I would like to get this test passing 100% of the time.
Also as a side note, I tested the old version of this test against the `master` branch and it works every time (test fails as there are bonus spins).

Reply:
The remaining issue that you were noticing is from the spinner potentially going away before the next check runs. So really you needed to keep a reference to the spinner and keep checking it directly. This actually goes one level deeper, as the `DrawableSpinner` may have been "unpooled", losing its judgement state and `HitObject` reference.

I've rewritten the test to take this into account, please take a look at the changes to understand how you can write these tests in the future.

Looking for a third party review to ensure the logic I've added is correctly.

Reply:
I think it looks OK, but [the latest test failures](https://github.com/ppy/osu/runs/5283951075) may indicate that this same method needs to be transplanted to `TestSpinnerAutoCompleted` or others.

Reply:
For sure. But I guess this was a failure that has been occurring until now, because nothing in this PR should have affected this...

Reply:
I've fixed that test, but left the remaining one as-is (doesn't seem to be failing due to checking earlier into the spinner... and is probably not possible to write in a better way without further consideration as it's checking the spinning speed specifically).